Remove temporary VarDecl from TapExpr

beccadax opened the pull request:
This PR redesigns `TapExpr` so that it doesn't need to declare a variable to hold the value it is modifying. Instead, it allocates a temporary in SILGen and uses an `OpaqueValueExpr` to represent references to the temporary. This allows us to remove various hacks, workarounds, and special cases.

To achieve this, the PR also modifies some of the invariants around OVEs and changes how ASTVerifier checks them. It formalizes the notion that OVEs belong to one of their ancestor nodes and adds calls to determine an OVE's owner, while removing the (increasingly untrue) invariant that OVEs don't appear in un-typechecked ASTs.

~~I hoped to remove some duplication in SILGen as well, but this turned out to be more difficult, so I'll create a separate PR with that change to discuss how to make it happen.~~ Finally, this removes some code duplication in SILGen which was added to work around limitations of the OSLogInterpolation pass.

None of the commits should change functionality; however, the new code patterns appear to be more optimizable.

brentdax commented:
@swift-ci please smoke test

slavapestov commented on lib/Sema/TypeCheckCaptures.cpp:
```diff
@@ -568,7 +560,7 @@ class FindCapturedVars : public ASTWalker {
     // Capture a placeholder opaque value.
     if (auto opaqueValue = dyn_cast<OpaqueValueExpr>(E)) {
-      if (opaqueValue->isPlaceholder()) {
+      if (opaqueValue->getOwnerKind() == ExprKind::Call /* FIXME */) {
```
It looks like this is the only time that getOwnerKind() is checked outside of an assertion. Perhaps the capture analysis can track which Exprs introduce which OVEs, and then record and "free" OVEs, instead?

brentdax replied:
Having gone through the exercise of doing this, there seem to currently be two interesting pieces of information: (1) whether this OVE can be captured and (2) whether we should ever see this OVE before CSApply. I could reduce it down to those two bits instead of storing the whole ExprKind (which doesn't quite work anyway).

Alternatively, I think the only problem with capturing OVEs is that Sema only knows how to analyzes a single OVE capture, even though we could represent arbitrary numbers of them as `CaptureValues`. If that's true, we could refactor to allow *n* OVE captures.

(One thing I noticed while working on this is that there are several different walks in the compiler which each separately have knowledge of which nodes can own OVEs. Is there some way we could centralize that knowledge? Maybe an `ASTWalker` subclass that automatically maintains OVE -> owner tables?)

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -5358,29 +5363,27 @@ RValue RValueEmitter::visitUnevaluatedInstanceExpr(UnevaluatedInstanceExpr *E,
+  auto VarAddress = VarInit->getManagedAddress();
+  SmallVector<std::unique_ptr<SILGenFunction::OpaqueValueRAII>, 5> opaqueValues;
+
+  for (auto use : E->getUses())
```
At one point SILGen allowed an OVE to be referenced multiple times. I wonder if it would be difficult to bring that back. We'd still want them to be unique in most situations, but TapExpr could be an exception.

brentdax replied:
A third "multiple use" bit, I suppose?

beccadax committed:
[NFC] Consistently chain up shouldVerify() and cleanup() calls

beccadax committed:
[NFC] OpaqueValueExpr::isPlaceholder() -> canBeCaptured()

This commit renames “isPlaceholder” to “canBeCaptured” and removes invariants stating that only “isPlaceholder” OVEs can appear in un-typechecked code. It turns out that we will want some “placeholder” OVEs to not be capturable, and the exact set of OVEs that should never appear in un-typechecked code is not a concept we need outside of assertions.

beccadax committed:
Remove the $interpolation variable

Previously, TapExpr used an actual, physical VarDecl to store the value it was modifying. That allowed us to represent references to that value as DeclRefExprs, which gave us some type-checking functionality for free but introduced issues like scoping that were not really necessary. This commit redesigns TapExpr to instead use OpaqueValueExprs as stand-ins for the value being modified. The backing storage is not directly represented in the AST; instead it is emitted in SILGen.

This commit changes two existing tests:

* DebugInfo/StringInterpolation.swift tested that the $interpolation variable was marked as artificial so it would be ignored. There is no such variable anymore, so it is now unnecessary.
* SILGen/sil_locations.swift had one SourceLoc move from inside the string literal to before it. This is fine.

IDE/complete_at_top_level.swift is broken in this commit, but will be fixed in a future commit.

beccadax committed:
[NFC] Formalize OpaqueValueExpr ownership

This commit directly expresses the rules about which OpaqueValueExprs belong to which ancestor nodes. It also ensures the ASTVerifier checks both that invariant and the invariant that each OVE is used exactly once. Finally, it enforces these invariants even in parsed but un-typechecked code.

beccadax committed:
[NFC] Allow certain OVEs to be reused

Nothing actually uses this yet.

beccadax committed:
Use a reusable OVE with TapExpr

This simplifies many aspects of TapExpr and fixes the last remaining test failure.

brentdax commented:
@swift-ci please smoke test

brentdax commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -2384,17 +2384,17 @@ visitInterpolatedStringLiteralExpr(InterpolatedStringLiteralExpr *E,
   {
     TapExpr *ETap = E->getAppendingExpr();
+
     // Inlined from TapExpr:
```
I will try to clean up this duplication tomorrow.

brentdax replied:
I was able to improve the factoring a little bit, but I had trouble actually removing the duplication because it reorders code in a way that disrupts the OSLogOptimization pass. I'll tag Ravi in a follow-up PR and see if we can figure something out.

beccadax committed:
Correctly clean interpolation OVE from the AST

SanitizeExpr was clearing an InterpolatedStringLiteralExpr’s InterpolationExpr (the OVE representing its builder instance) from its TapExpr’s SubExpr, but not from the ISLE itself. This could cause AST validation failures in code in invalid code because the OVE would be introduced by the ISLE but would not actually exist in the AST.

This commit:
* Improves SanitizeExpr to clean out both references to the OVE.
* Tweaks ASTVerifier to account for the fact that the InterpolationExpr may not exist.
* Adds InterpolationExpr to AST dumps to assist in reading interpolated string literals in them.

brentdax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Factor helper func from interpolation SILGen

This helps clarify the structure of visitInterpolatedStringLiteralExpr() to aid in future refactoring.

brentdax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Refactor SILGen for interpolations

brentdax commented:
@swift-ci please smoke test

beccadax committed:
[SILGen] [NFC] Remove interpolation duplication

Previous refactorings have gotten us to the point where we can now delegate to `visitTapExpr()` again instead of emitting the TapExpr specially in `RValueEmitter::visitInterpolatedStringLiteralExpr()`.

brentdax commented:
Ravi's change in #27419 has gotten this to the point where I can actually finish the refactoring and remove the duplication in SILGen. 🎉

brentdax commented:
I'm expecting one failure in NameBinding/name_lookup.swift, but we can see how the rest of it goes.

@swift-ci please smoke test

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please smoke test

pschuh commented on include/swift/AST/Expr.h:
```diff
@@ -3853,15 +3862,66 @@ class OpaqueValueExpr : public Expr {
+
+  /// Whether the opaque value can only appear in one location in the AST or in
+  /// several.
+  bool canBeReused() const {
```
Should this be used somewhere in SILGen, or is it only for validation?

brentdax replied:
I only ended up needing it in validation.

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
